### PR TITLE
feat: add index options to many-to-one relation decorator

### LINKED ISSUE #7118: Add index option to ManyToOne decorator
## Feature Description

Add a quality-of-life indexing option to ManyToOne decorator.

### The Problem

Not all RDBMSs auto-index the referencing-side of FKs (postgresql doesn't do it), and very frequently there are benefits of doing so. Since this ought to be a common idiom it makes sense for the ManyToOne decorator to accept an option which tell it to also create an index for the FK.

### The Solution

Add a `index?: boolean | IndexOptions & { name?: string }` property to ManyToOneRelations(a new interface which extends `RelationOptions`) which the decorator reads and creates the index metadata if it is set.

### Relevant Database Driver(s)

- [ ] `aurora-data-api`
- [ ] `aurora-data-api-pg`
- [ ] `better-sqlite3`
- [ ] `cockroachdb`
- [ ] `cordova`
- [ ] `expo`
- [ ] `mongodb`
- [ ] `mysql`
- [ ] `nativescript`
- [ ] `oracle`
- [x] `postgres`
- [ ] `react-native`
- [ ] `sap`
- [ ] `sqlite`
- [ ] `sqlite-abstract`
- [ ] `sqljs`
- [ ] `sqlserver`

### Are you willing to resolve this issue by submitting a Pull Request?

- [x] Yes, I have the time, and I know how to start.
- [ ] Yes, I have the time, but I don't know how to start. I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

### PR DESCRIPTION
**Pull request by pedrolcn**
### Description of change

Adds option to create index on the referencing side of Many-To-One relations.  
Closes #7118 

### Pull-Request Checklist

- [ ] Code is up-to-date with the `master` branch
- [ ] `npm run lint` passes with this change
- [ ] `npm run test` passes with this change
- [ ] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [ ] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/decorator/options/ManyToOneOptions.ts
Comment (gabrielsanttana) [inline src/decorator/options/ManyToOneOptions.ts:4]: Here you could remove this white space:

```suggestion
```

File: src/decorator/relations/ManyToOne.ts
Comment (nebkat) [inline src/decorator/relations/ManyToOne.ts:75]: I suggest you reuse the existing `Index()` decorator here so we don't duplicate.
```suggestion
            Index(indexOptions)(object, propertyName);
```

**Review by nebkat — CHANGES_REQUESTED** [review]

File: src/decorator/relations/ManyToOne.ts
Comment (pedrolcn) [reply src/decorator/relations/ManyToOne.ts]: Done. Also added logic for handling custom index names without code duplication.

File: src/decorator/options/ManyToOneOptions.ts
Comment (pedrolcn) [reply src/decorator/options/ManyToOneOptions.ts:3]: Done

**Comment by nebkat** [comment]
Looks good, last thing just squash these 4 commits into one for less git spam :tada:

**Comment by pedrolcn** [comment]
> Looks good, last thing just squash these 4 commits into one for less git spam 🎉

Done!

**Comment by pedrolcn** [comment]
@nebkat 

This has been open for a few days, is there any action from me that is still needed before this can be merged ?

**Review by nebkat — APPROVED** [review]
@pedrolcn I am not a maintainer here, just helping out. There haven't been any merges in over a month so just a matter of waiting now. Approved by me anyway!

**Comment by pedrolcn** [comment]
> @pedrolcn I am not a maintainer here, just helping out. There haven't been any merges in over a month so just a matter of waiting now. Approved by me anyway!

Oh, I didn't realize that, I see
Thanks for the response!

**Comment by pleerock** [comment]
Thank you for your efforts!

However, why do we need this feature when we already have a `Index` decorator?
What you are suggesting is to introduce this:

```ts
@ManyToOne(() => Type, type => type.property, {
   option1: ... 
})
```

when we already have:

```ts
@Index({ option1: ... })
@ManyToOne(() => Type, type => type.property)
```

Don't see much of point. Next step would be to merge `@JoinColumn` into this decorator or even worse `@JoinTable`.

From my experience this kind of features only confuses people and complicates the code because of new flows in the code, e.g.:

* what if I put both `@Index({ someOptions })` and `@Index({ someOptions }) @ManyToOne({ index: otherOptions })`?
* user can ask what is the difference between putting index into `many-to-one` and separate index, needs additional documentation

**Comment by nebkat** [comment]
But couldn't we then say the same for `@Column({ generated: true, unique: true, primary: true})` and every combination of `@Generated()`  `@Unique()` `@PrimaryColumn()` `@PrimaryGeneratedColumn()`? In both cases it's a choice between being explicit or concise, or even ultimately a matter of style. Personally I prefer to use `@Unique()`/`@Index()` only for multi-column indices.

If you really want to solve the user confusion with these sort of things I think we need to move towards separate documentation and auto generated API reference. The user should be able to look up `ColumnOptions` and see that `generated: true` is the same as `Generated()` and vice-versa, rather than someone having to manually document the little details (of which there are many) in `docs/`. It also wouldn't be all that difficult to have all the `MetadataArgsStorage` additions check if there are duplicate definitions and throw a warning or error.

----

What I've tried to do here https://github.com/typeorm/typeorm/pull/7058/commits/b032f6b0ef9be87fbd4ad7813a382519f9ada11a https://github.com/typeorm/typeorm/pull/7058/commits/63c9bd95eb82e5c55e73c06b392e0c520a519c1d https://github.com/typeorm/typeorm/pull/7058/commits/b321c020be76f1a90647728e5c939830a66474ef is at least have all the duplicate functionality use the same code by calling decorators within decorators. As long as this is the case I don't see any harm in having both `@Column({option: ...})` and `@Option`.

----

Also slight correction just in case you misread rather than made a typo, it would be:

```ts
@ManyToOne(() => Type, type => type.property, {
   index: {option1: ... }
})
```

Rather than:

```ts
@ManyToOne(() => Type, type => type.property, {
   option1: ... 
})
```

Which is a bit more obvious as to what is happening.

**Comment by pedrolcn** [comment]
@pleerock 
The rationale is that indexing a FK is a common enough idiom (and most likely a best practice) that it should be encouraged by the API design as well something that should be very ergonomic to do. But as the linked issue mentions, this is mostly a quality-of-life improvement, it is not by any chance something that was not possible before.

Although, I do agree that "many ways of doing the same thing" is a problem and one that TypeORM definitely suffers from (sometimes with subtle differences between them) and it can confuse users. I myself haven't stopped to consider what happens if the user does something like
```typescript
@Index({/* indexOptions */})
@ManyToOne(() => Type, type => type.property, {
  index: { /* indexOptions */  }
})
```

**Review by leonidasv — APPROVED** [review]
Nice work.

**Comment by pleerock** [comment]
> If you really want to solve the user confusion with these sort of things I think we need to move towards separate documentation and auto generated API reference. 

> I do agree that "many ways of doing the same thing" is a problem and one that TypeORM definitely suffers from (sometimes with subtle differences between them) and it can confuse users

I think you both admit that current design is confusing and I don't really want to continue with new features that bring a new confusion. Let me close it for now, until we have time to redesign current confusing things and make a better design.

**Pull request closed without merging**
